### ISSUE TITLE
flux gitlab does not work with subgroups

### ISSUE THREAD

**Issue by schmidp** [fluxcd/flux2#3869]
### Describe the bug

GitLab requires API calls to subgroups to encode the forward slash between the group and the subgroup:

https://docs.gitlab.com/ee/api/rest/index.html#namespaced-path-encoding

flux is not adhering to this requirement.

### Steps to reproduce

1. run the following flux command:
```
flux bootstrap gitlab --hostname=hostname--token-auth --owner=group/subgroup --repository=repo --branch=main --path=clusters/my-cluster
```
2. You will get the following result:
```
Output: ►
│ connecting to https://hostname/
│ ✗ failed to create new Git repository
│ "https://hostname/group/subgroup/repo":
│ GET
│ https://hostname/group/subgroup/repo:
│ 404 {message: 404 Group Not Found}
```

If you try with `group%2Fsubgroup`:
1. run the following flux command:
```
flux bootstrap gitlab --hostname=hostname--token-auth --owner=group%2Fsubgroup --repository=repo --branch=main --path=clusters/my-cluster
```
2. You will get the following result:
```
Output: ►
│ connecting to https://hostname/
│ ✗ failed to create new Git repository
│ "https://hostname/group%2Fsubgroup/repo":
│ GET
│ https://hostname/api/v4/groups/group/subgroup:
│ 404 {message: 404 Group Not Found}
```

You can also try via the command line:

```
curl --header "Private-Token: XXXXX" "https://hostname/api/v4/projects/group/subgroup/repo"
{"error":"404 Not Found"}
```

vs:

```
curl --header "Private-Token: XXXXX" "https://hostname/api/v4/projects/group%2Fsubgroup%2Frepo"
{"id":123,"description":null,"name":"repo...
```

### Expected behavior

Flux encodes the owner correctly so that the api called can be completed.

### OS / Distro

N/A

### Flux version

flux version 2.0.0-rc.2

### Flux check

N/A

### Git provider

GitLab

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by stefanprodan** [comment]
Can you please try `flux bootstrap git --url` and see if this works with subgroups? You'll need to set the deploy key by hand but it should unblock you until we figure out a fix.

**Comment by somtochiama** [comment]
@schmidp can you verify that what you have set as the group is the same as the group URL (you can check the URL when you visit the group's page in the browser)

When you create a group, the id given to the group might not be the same as its name and this isn't shown in the UI when you go to the group
![Screenshot 2023-05-12 at 15 24 38](https://github.com/fluxcd/flux2/assets/34774756/a7687cb2-1825-47ed-8af1-b46df4f07096)

Can you try this curl command and see if it works:
```
curl --header "Private-Token: XXXXX" "https://hostname/api/v4/groups%2Fgroup%2Fsubgroup
```

**Comment by schmidp** [comment]
@somtochiama yes it is definitely the same one (I made sure by checking "settings/advanced/Change group URL".

I have tried your command and it leads to a 404. If I replace the %2F after groups with a / it works.
If I also replace the %2F between group and subgroup I get a 404 again.

@stefanprodan: thanks. I have already switched to the terraform provider and am a very happy user of flux now :)

**Comment by somtochiama** [comment]
> I have tried your command and it leads to a 404. If I replace the %2F after groups with a / it works.
Yes, this is the correct URL, I made a mistake in my comment

I have been unable to recreate your issue. I only get this error if I specify the wrong group id or entered an empty token(forgot to copy it first) at the prompt. Maybe it's an issue with self-hosted Gitlab. I will deploy one and see if I get the error. 

Can you try to bootstrap(on a test cluster) with another group? Do you still get the same error? Export the token using `export GITLAB_TOKEN=XXX` so the prompt doesn't come up.

**Comment by somtochiama** [comment]
I tried with self-hosted Gitlab and I wasn't able to reproduce this

**Comment by schmidp** [comment]
Very strange. Especially since the docs also state *to make sure* to encode the paths: https://docs.gitlab.com/ee/api/rest/index.html#namespaced-path-encoding

We are using an GitLab Omnibus self-hosted instance. The instance is many years old maybe something changed.
I am on holiday from tomorrow on, but if I have time when I am back I will try to spin up an Omnibus self-hosted instance and see if it behaves differently from ours. But still find it strange that the docs say to use path encoding but it works without for you.

**Comment by somtochiama** [comment]
> But still find it strange that the docs say to use path encoding but it works without for you.

The path encoding actually happens under the hood, The  CLI displays a different URL from what was actually used because that's what the library returns (This is something we should work on so it's not confusing).

**Comment by romanteres** [comment]
I had similar issue using service account. I resolved the issue by adding the account as member into my organiosation group.

https://docs.gitlab.com/user/profile/service_accounts/

documentation says

"Service account access to groups and projects is managed the same way as human users in the UI. For more information, see groups and members of a project.

You can assign service accounts to groups and projects using the UI or the members API."
